[CI] DeepSeek multi-host module tests: sampling trace capture hits remote-device TT_FATAL on DUAL

## Summary
A new DeepSeek multi-host failure family is now surfacing on `main` in the DUAL module job. During device-sampling trace capture, the job fatals with:

- `TT_FATAL: Cannot get device for remote device at coordinate MeshCoordinate([3, 1])`
- `TT_FATAL: Cannot get device for remote device at coordinate MeshCoordinate([7, 7])`

This is distinct from the older multi-host failures already tracked in:

- #42014 (`test_bspm_demo.py` shared-cache `PermissionError`)
- #37051 (stale file handle / shared-drive failures)

## Latest evidence
- April 15, 2026 `main` run [`#349`](https://github.com/tenstorrent/tt-metal/actions/runs/24437410983), job [`deepseekv3-tests (dual)`](https://github.com/tenstorrent/tt-metal/actions/runs/24437410983/job/71395167688)
  - log excerpt around lines 16744-16748:
    - `models.common.sampling.generator:capture_trace:284 - Pre-compiling sampling path before trace capture...`
    - `TT_FATAL: Cannot get device for remote device at coordinate MeshCoordinate([3, 1])`
    - `TT_FATAL: Cannot get device for remote device at coordinate MeshCoordinate([7, 7])`
    - `##[error]The action has timed out.`
- April 14, 2026 first observed `main` run [`#339`](https://github.com/tenstorrent/tt-metal/actions/runs/24381918919), job [`deepseekv3-tests (dual)`](https://github.com/tenstorrent/tt-metal/actions/runs/24381918919/job/71207828690)
  - same failure around lines 27780-27784 with the same test case and the same two mesh coordinates.

## Failure point
- failing test:
  - `models/demos/deepseek_v3/tests/test_sampling.py::test_deepseek_device_sampling_argmax_path[device_params0-sampling_params2]`
- trace capture path:
  - [`models/common/sampling/generator.py#L284-L300`](https://github.com/tenstorrent/tt-metal/blob/main/models/common/sampling/generator.py#L284-L300)
- the fatal happens immediately after `capture_trace()` precompile logging and before the trace completes.

## Why this looks new
- The April 13, 2026 `main` run [`#335`](https://github.com/tenstorrent/tt-metal/actions/runs/24326790870) failed earlier in `test_bspm_demo.py` with the separate shared-cache `PermissionError` tracked by #42014, so this path was not reached there.
- April 14, 2026 `main` run `#339` is the first completed `main` run that reaches this sampling trace-capture path and shows the remote-device `TT_FATAL`.
- April 15, 2026 `main` run `#349` reproduces the exact same signature on a newer head SHA, so this is not a one-off flake.

## Smallest observable introduction window
- last `main` head before this signature surfaced in CI: `1ed365693c8f490ad21e3e351f2f2709db4b2019` (run `#335`, April 13, 2026)  
  note: that run failed earlier for the separate BSPM cache issue, so this signature may have been latent but masked.
- first `main` head with this signature: [`90c914ef258b5cc92ad172f3604b784ec77253ca`](https://github.com/tenstorrent/tt-metal/commit/90c914ef258b5cc92ad172f3604b784ec77253ca) (run `#339`, April 14, 2026)
- compare view:
  - https://github.com/tenstorrent/tt-metal/compare/1ed365693c8f490ad21e3e351f2f2709db4b2019...90c914ef258b5cc92ad172f3604b784ec77253ca

Among the commits in that window, the only one that directly rewrites distributed mesh trace capture / dispatch is [`#41094`](https://github.com/tenstorrent/tt-metal/pull/41094) / `90c914ef258b5cc92ad172f3604b784ec77253ca` (`Optimize mesh trace dispatch by skipping redundant kernel binary writes`). The other relevant-window changes were a CI harness adjustment, a teacher-force test-only fix, and a DeepSeek MoE integration change, none of which line up with this trace-capture failure site.

## Likely root cause
`#41094` removed the old local-subgrid-only trace capture flow and replaced it with deferred `TraceNode` assembly.

Relevant code points:
- new trace recording stores trace nodes for every coordinate/range during bypass-mode recording, without the old local-subgrid filtering:
  - [`tt_metal/distributed/fd_mesh_command_queue.cpp@90c914e#L310-L334`](https://github.com/tenstorrent/tt-metal/blob/90c914ef258b5cc92ad172f3604b784ec77253ca/tt_metal/distributed/fd_mesh_command_queue.cpp#L310-L334)
- during `record_end`, the new path patches per-device launch messages via `mesh_device_->get_device(range.start_coord())`:
  - [`tt_metal/distributed/fd_mesh_command_queue.cpp@90c914e#L1338-L1343`](https://github.com/tenstorrent/tt-metal/blob/90c914ef258b5cc92ad172f3604b784ec77253ca/tt_metal/distributed/fd_mesh_command_queue.cpp#L1338-L1343)
- `MeshDeviceViewImpl::get_device()` explicitly fatals on remote coordinates:
  - [`tt_metal/distributed/mesh_device_view.cpp#L172-L178`](https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/distributed/mesh_device_view.cpp#L172-L178)

The old implementation explicitly intersected the program range with the local mesh view and skipped unused remote subgrids before touching device-local state. The new deferred trace path appears to have dropped that protection. On DUAL multi-host sampling trace capture, that leaves `record_end()` attempting to resolve a remote `range.start_coord()` as a local device and aborting.

## Likely owner
- likely introducing contributor: @jbaumanTT (author of `#41094` / `90c914ef258b5cc92ad172f3604b784ec77253ca`)
- CODEOWNERS for `tt_metal/distributed/`: `@cfjchu @aliuTT @tt-asaigal @jbaumanTT`

## Suggested next step
- audit the `#41094` trace rewrite for places where multi-host trace assembly now uses raw mesh coordinates without first restricting to the local partition / local mesh view.
- in particular, the `record_end()` path that patches Tracy launch messages should likely not call `get_device()` on a remote `range.start_coord()`.

--- a/tt_metal/distributed/fd_mesh_command_queue.cpp
+++ b/tt_metal/distributed/fd_mesh_command_queue.cpp
@@ -1099,21 +1099,27 @@ void FDMeshCommandQueue::record_end() {
     // trace to ensure they are equal at the end.
 
     // Calculate device ranges that have an identical set of programs that run on them (including no programs at all).
-    std::vector<MeshCoordinateRange> device_ranges{MeshCoordinateRange{mesh_device_->shape()}};
+    // Restrict to the local mesh partition so that multi-host traces only process locally-owned devices.
+    auto local_mesh_range = mesh_device_->get_view().get_local_mesh_coord_range();
+    std::vector<MeshCoordinateRange> device_ranges{local_mesh_range};
     for (auto& trace_node : trace_nodes_) {
         for (auto& [device_range, program] : trace_node.trace_nodes) {
+            auto local_device_range = local_mesh_range.intersection(device_range);
+            if (!local_device_range.has_value()) {
+                continue;
+            }
             bool intersection_found = false;
             std::vector<size_t> device_range_idxs_to_invalidate;
             for (size_t i = 0; i < device_ranges.size(); i++) {
                 auto& existing_range = device_ranges[i];
                 TT_FATAL(
-                    existing_range.dims() == device_range.dims(),
+                    existing_range.dims() == local_device_range->dims(),
                     "Invalid mismatching dimensions for existing {} vs device range {}",
                     existing_range.dims(),
-                    device_range.dims());
-                if (existing_range.intersects(device_range)) {
+                    local_device_range->dims());
+                if (existing_range.intersects(*local_device_range)) {
                     intersection_found = true;
-                    auto intersection = *existing_range.intersection(device_range);
+                    auto intersection = *existing_range.intersection(*local_device_range);
                     if (intersection != existing_range) {
                         auto complement = subtract(existing_range, intersection);
                         device_range_idxs_to_invalidate.push_back(i);
@@ -1135,7 +1141,7 @@ void FDMeshCommandQueue::record_end() {
                         device_ranges.end());
                 }
             } else {
-                device_ranges.push_back(device_range);
+                device_ranges.push_back(*local_device_range);
             }
         }
     }

